### DISCUSSION TITLE
Android TV, Channels on the home screen

### DISCUSSION THREAD

**Discussion by firemen01** [DelphiWorlds/Kastri#241] [General]
Does anyone have experience working with channels on the main screen of Android TV? show how it's done  in Delphi. I'm not very good with Java.

**Comment by DelphiWorlds** [comment]
Please give more detail for what you mean by "working with channels". If you have links to Java code that do what you want to achieve, please provide them. Some assistance with converting to Delphi code might be given here

**Comment by DelphiWorlds** [comment]
I decided anyway to do a little research into what you might mean. Are you referring to the [TvInputManager](https://developer.android.com/reference/android/media/tv/TvInputManager) and associated classes? It might help to know exactly what you want to do with it. Again, if you have links to Java examples, please provide them, indicating which parts are relevant.

**Comment by firemen01** [comment]
I meant it:  
https://developer.android.com/training/tv/discovery/recommendations-channel

git: 
https://github.com/android/tv-samples/tree/main/ClassicsKotlin
(TV launcher home screen channels)

**Comment by DelphiWorlds** [comment]
This still does not tell me exactly what you want to achieve. That Kotlin example is quite extensive - if you want a complete Delphi equivalent it would take some time, and would be quite costly if you hired me to do it.

**Comment by firemen01** [comment]
I just need to be able to place cards on the home screen in my app's feed. how YouTube shows video recommendations on the main screen.
Don't you have a device with Android TV?
Everything is clearly described here with examples in Java:
https://developer.android.com/training/tv/discovery/recommendations-channel
It seems like everything is simple and just a couple of lines of code, but I don’t know how to approach it. Maybe your dear visitors have implemented this in Delphi and can help?

**Comment by DelphiWorlds** [comment]
> Don't you have a device with Android TV?

I do have an Android TV, as well as being able to use an emulator

> I just need to be able to place cards on the home screen in my app's feed. how YouTube shows video recommendations on the main screen.

OK.. that gives me a better idea, thanks

**Comment by DelphiWorlds** [comment]
I've done the first step, which is import the [`tvprovider-1.0.0.jar`](https://github.com/DelphiWorlds/Kastri/blob/master/ThirdParty/Android/tvprovider-1.0.0.jar) into a [Delphi unit](https://github.com/DelphiWorlds/Kastri/blob/master/API/DW.Androidapi.JNI.AndroidX.TvProvider.pas)

**Comment by firemen01** [comment]
Thank you!
Previously, I took the same tvprovider-1.0.0.jar and got the wrapper via Java2OP.exe -jar tvprovider-1.0.0.jar.
but so far nothing has worked out in terms of displaying my channels

**Comment by DelphiWorlds** [comment]
> so far nothing has worked out in terms of displaying my channels

Does it add a channel at all? Can you show the code that you tried?

**Comment by firemen01** [comment]
Tell me, what emulator are you using, where is there full support for Android TV?
I abandoned and deleted everything, because nothing worked and it was inconvenient to test on the TV, and my wife was constantly grumbling)) I have a BlueStacks emulator, but I don’t understand how to install Android TV Core Services

**Comment by DelphiWorlds** [comment]
> Tell me, what emulator are you using, where is there full support for Android TV?

The emulator I use is from Android Studio, however I am running Android Studio on an M1 Mac (arm-based), and use an SSH tunnel to be able to set up a connection between Delphi and the emulator, which uses an arm64 image. The emulators in Android Studio for non-arm images are way too slow to work with, so that would probably not be suitable for you.

> I abandoned and deleted everything..

You mean you deleted all your code?

**Comment by firemen01** [comment]
I poked around a little and settled on this for now:

```
 <uses-permission android:name="com.android.providers.tv.permission.WRITE_EPG_DATA" />

DW.Androidapi.JNI.AndroidX.TvProvider,  Androidapi.Helpers, AndroidApi.JNI.Net, Androidapi.JNI.GraphicsContentViewText,  FMX.Surfaces, FMX.Helpers.Android,

function AddProgramToChannel(channelId: Int64; const title, description: string; const uri: Jnet_Uri; const appProgramId: string): Jnet_Uri;
var
  Context: JContext;
  Builder: JPreviewProgram_Builder;
  ProgramUri: Jnet_Uri;
begin
  Context := TAndroidHelper.Context;

  Builder := TJPreviewProgram_Builder.JavaClass.init;
  Builder.setChannelId(channelId).setType(TJTvContractCompat_PreviewPrograms.JavaClass.TYPE_CLIP);
  Builder.setChannelId(channelId).setTitle(StringToJString(title));
  Builder.setChannelId(channelId).setDescription(StringToJString(description));
  Builder.setChannelId(channelId).setPosterArtUri(uri);
  Builder.setChannelId(channelId).setIntentUri(uri);
  Builder.setChannelId(channelId).setInternalProviderId(StringToJString(appProgramId));

  ProgramUri := Context.getContentResolver().insert(TJTvContractCompat_PreviewPrograms.JavaClass.CONTENT_URI, Builder.build().toContentValues());

  result:= ProgramUri;
end;

function BitmapToJBitmap(const ABitmap: TBitmap): JBitmap;
var
  LSurface: TBitmapSurface;
begin
  Result := TJBitmap.JavaClass.createBitmap(ABitmap.Width, ABitmap.Height, TJBitmap_Config.JavaClass.ARGB_8888);
  LSurface := TBitmapSurface.Create;
  try
    LSurface.Assign(ABitmap);
    SurfaceToJBitmap(LSurface, Result);
  finally
    LSurface.Free;
  end;
end;

function StoreChannelLogo(channelId: Int64; const logoUri: string; const logoBitmap: TBitmap): boolean;
var
  LogoUriJ: Jnet_Uri;
  JBitmap:  Androidapi.JNI.GraphicsContentViewText.JBitmap;
  Success: Boolean;
begin

  if (logoBitmap <> nil) then begin
    JBitmap := BitmapToJBitmap(logoBitmap);
    Success := TJChannelLogoUtils.JavaClass.storeChannelLogo(TAndroidHelper.Context, channelId, JBitmap);
  end else begin
    LogoUriJ := TJnet_Uri.JavaClass.parse(StringToJString(logoUri));
    Success := TJChannelLogoUtils.JavaClass.storeChannelLogo(TAndroidHelper.Context, channelId, LogoUriJ);
  end;

  result:= Success;

end;

procedure TForm1.Button1Click(Sender: TObject);
var
   ChannelUri,
   AppLinkUri: Jnet_Uri;
   ContentResolver: JContentResolver;
   ChannelId: Int64;
  ChannelBuilder: JChannel_Builder;
begin
   ChannelBuilder:= TJChannel_Builder.Create;

   ChannelBuilder.setType(TJTvContractCompat_Channels.JavaClass.TYPE_PREVIEW);
   ChannelBuilder.setDisplayName(StringToJString('myDisplayName'));
   ChannelBuilder.setDescription(StringToJString('myTested Channel'));

   AppLinkUri := TJnet_Uri.JavaClass.parse(StringToJString('https://google.com'));

   ChannelBuilder.setAppLinkIntentUri(AppLinkUri);

   ContentResolver := SharedActivityContext.getContentResolver;

   ChannelUri := ContentResolver.insert(TJTvContractCompat_Channels.JavaClass.CONTENT_URI,  ChannelBuilder.build().toContentValues() );

  if ChannelUri <> nil then begin

    Edit1.Text:= JStringToString(ChannelUri.toString);
    ChannelId := TJContentUris.JavaClass.parseId(ChannelUri);

    Edit2.Text:= IntToStr(ChannelId);

    TJTvContractCompat.JavaClass.requestChannelBrowsable(TAndroidHelper.Context, channelId);

    var Bitmap: TBitmap;
    Bitmap := TBitmap.Create(80, 80);
    try
      Bitmap.Clear(TAlphaColorRec.Red);
      StoreChannelLogo(channelId, '', Bitmap);
    finally
      Bitmap.free;
    end;

  end else Edit1.Text:= 'Error';

end;
```

I get ChannelId normally.

**Comment by DelphiWorlds** [comment]
The code all looks like it should work - is the channel not appearing?

**Comment by DelphiWorlds** [comment]
I've done a test myself, and there was a problem with the import unit, which I have [now updated](https://github.com/DelphiWorlds/Kastri/blob/master/API/DW.Androidapi.JNI.AndroidX.TvProvider.pas).

I was unable to insert a program, though - it throws this error (in the logcat messages)

```
08-12 08:56:58.833	com.android.providers.tv	1733	12195	E	SQLiteDatabase	 Error inserting package_name=com.embarcadero.ATVDemo internal_provider_id=12345 type=0	
08-12 08:56:58.833	com.android.providers.tv	1733	12195	E	SQLiteDatabase	 android.database.sqlite.SQLiteConstraintException: NOT NULL constraint failed: channels.input_id (code 1299 SQLITE_CONSTRAINT_NOTNULL)	

```

You might have more luck...

**Comment by firemen01** [comment]
I added it to the manifest:
```
        </activity>
		
        <meta-data
        android:name="android.media.tv.INPUT_ID"
        android:value="com.embarcadero.ATVDemo" />	
```
I also indicated the channel ID in the code:
```

ChannelBuilder.setInputId(StringToJString('com.embarcadero.ATVDemo'));

...........

var ProgramUri: Jnet_Uri;
ProgramUri := AddProgramToChannel(channelId, 'myProgramm', 'Description', ChannelUri, AppLinkUri, 'myProgramId1'); 

..........

function AddProgramToChannel(channelId: Int64; const title, description: string; const ChannelUri, AppUri: Jnet_Uri; const appProgramId: string): Jnet_Uri;
var
  Context: JContext;
  Builder: JPreviewProgram_Builder;
  ProgramUri: Jnet_Uri;
begin
  Context := TAndroidHelper.Context;

  Builder := TJPreviewProgram_Builder.JavaClass.init;
  Builder.setChannelId(channelId).setType(TJTvContractCompat_PreviewPrograms.JavaClass.TYPE_CLIP);
  Builder.setChannelId(channelId).setTitle(StringToJString(title));
  Builder.setChannelId(channelId).setDescription(StringToJString(description));
  Builder.setChannelId(channelId).setPosterArtUri(ChannelUri);
  Builder.setChannelId(channelId).setIntentUri(AppUri);
  Builder.setChannelId(channelId).setInternalProviderId(StringToJString(appProgramId));

  Builder.setChannelId(channelId).setInternalProviderId(StringToJString(appProgramId));

  ProgramUri := Context.getContentResolver().insert(TJTvContractCompat_PreviewPrograms.JavaClass.CONTENT_URI, 
  Builder.build().toContentValues());

  result:= ProgramUri;
end;
```

as a result I get a ProgramUri but still nothing is displayed on the screen (:

**Comment by DelphiWorlds** [comment]
Even with using setInputId on the channel builder, I get the same error as before when inserting a preview program. I might copy your code and play around with it...

**Comment by DelphiWorlds** [comment]
I worked out what my problem was - I was using `TJTvContractCompat_Channels.JavaClass.CONTENT_URI` when inserting the program.. whoops!

I've now published a [demo in the HowTo repo](https://github.com/DelphiWorlds/HowTo/tree/main/Demos/AndroidTV) of my work so far.

**Comment by firemen01** [comment]
Thank you!
